File: cs/ant_system/AntSystem.cs
using System;
using System.Collections.Generic;
using System.IO;

class AntSystem{
    private double[,] distanceArr, distanceArrInv, pheromoneArr;
    private double alpha, beta, rho, pheromoneQ;

    public double bestFitnss = -1;
    public Agent bestAgent;
    private int cityNum;

    public AntSystem(int cityNum, string filename, double initPheromone, double alpha, double beta, double rho, double pheromoneQ)
    {
        this.cityNum = cityNum;
        this.alpha = alpha;
        this.beta = beta;
        this.rho = rho;
        this.pheromoneQ = pheromoneQ;
        distanceArr = new double[cityNum, cityNum];
        distanceArrInv = new double[cityNum, cityNum];
        makePheromoneArr(initPheromone);
        makeDistanceArr(cityNum, filename);
    }

    private void makePheromoneArr(double initPheromone){
        pheromoneArr = new double[cityNum, cityNum];
        for (int i=0; i<cityNum; i++){
            for (int j=i+1; j<cityNum; j++){
                pheromoneArr[i, j] = initPheromone;
                pheromoneArr[j, i] = initPheromone;
            }
        }
    }

    private void makeDistanceArr(int cityNum, string filename){
        double[,] cityInfo = new double[cityNum, 2];
        double dis;
        int count = 0, i, j;

        StreamReader sr = new StreamReader(filename);
        while(!sr.EndOfStream){
            string line = sr.ReadLine();
            string[] city = line.Split(",");
            cityInfo[count, 0] = double.Parse(city[0]);
            cityInfo[count, 1] = double.Parse(city[1]);
            count += 1;
        }
        sr.Close();

        for(i=0; i<cityNum; i++)
        {
            for(j=i+1; j<cityNum; j++){
                dis = Math.Pow(Math.Pow(cityInfo[i, 0] - cityInfo[j, 0], 2.0) + Math.Pow(cityInfo[i, 1] - cityInfo[j, 1], 2.0), 0.5);
                 distanceArr[i, j] = distanceArr[j, i] = dis;
                 distanceArrInv[i, j] = distanceArrInv[j, i] = dis;
            }
        }
    }

    public void generateRoute(Agent agent){
        double probDenominator, dRandom, probSum, distance=0.0;
        Random cRandom = new System.Random();
        Random firstCity = new System.Random();
        int fCity = firstCity.Next(0, cityNum);
        int i, j, preCity = fCity;

        agent.addRoute(preCity);
        for(i=1; i<cityNum; i++){
            double[] v = new double[cityNum];
            probDenominator = 0.0;

            for(j=0; j<cityNum; j++){
                if(!agent.isAlreadySet(j)){
                    double val = Math.Pow(pheromoneArr[preCity, j], alpha) * Math.Pow(distanceArrInv[preCity, j], beta);
                    v[j] = val;
                    probDenominator += val;
                }
            }
            dRandom = cRandom.NextDouble();
            probSum = 0;

            for(j=0; j<cityNum; j++){
                if(v[j] != 0){
                    probSum += v[j] / probDenominator;
                    if(probSum>dRandom){
                        agent.addRoute(j);
                        distance += distanceArr[preCity, j];
                        preCity = j;
                        break;
                    }
                }
            }
        }
        distance += distanceArr[preCity, fCity];
        agent.setDistance(distance);

        if(bestFitnss==-1 || distance < bestFitnss){
            bestFitnss = distance;
            bestAgent = agent;
        }
    }

    private void reducePheromone(){
        for(int i=0; i<cityNum; i++){
            for(int j=i+1; j<cityNum; j++){
                pheromoneArr[i, j] = pheromoneArr[i, j] * rho;
                pheromoneArr[j, i] = pheromoneArr[i, j];
            }
        }
    }

    public void updatePheromone(Agent agent){
        reducePheromone();
        double add = pheromoneQ / agent.distance;
        int city1, city2;
        for(int i=0; i<cityNum; i++){
            city1 = agent.route[i];
            city2 = agent.route[(i+1)%cityNum];
            pheromoneArr[city1, city2] += add;
            pheromoneArr[city2, city1] += add;
        }
    }
}

File: cpp/ant_system/ant_system.cpp
#include "ant_system.hpp"
#include "config.hpp"
#include "utils.hpp"
#include <iostream>
#include <ctime>
#include <cstdlib>
#include <cstdio>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>
#include <cmath>

using namespace std;

vector<string> split(string& input, char delimiter)
{
    istringstream stream(input);
    string field;
    vector<string> result;
    while (getline(stream, field, delimiter)) {
        result.push_back(field);
    }
    return result;
}


/******************************************************************************************/
Agent::Agent(){
    this->route = new int[CITY_NUM];
    for(int i=0; i<CITY_NUM; i++)    this->route[i] = 0;
}
 
Agent::~Agent(){
    delete[] this->route;
}

void Agent::reset_info(void){
    this->length = 0;
    this->registered_city_num = 1;
    for(int i=0; i<CITY_NUM; i++)    this->route[i] = 0;
    this->route[0] = this->id;
    this->last_city = this->id;
}


/******************************************************************************************/
AntSystem::AntSystem(){
    float dis;
    int dx, dy;
    int i, j;
    float** city_info = new float*[CITY_NUM];
    int count = 0;

    this->agent = new Agent[AGENT_NUM];
    for(i=0; i< AGENT_NUM; i++)  this->agent[i].set_id(i);
    for(i=0; i<CITY_NUM; i++){
        this->distance_arr[i] = new float[CITY_NUM];
        this->pheromone_arr[i] = new float[CITY_NUM];
        city_info[i] = new float[2];
    }

    ifstream ifs(get_benchmark());
    string line;
    
    while(getline(ifs, line)){
        vector<string> strvec = split(line, ',');
        for(i=0; i<strvec.size(); i++)  city_info[count][i] = stoi(strvec.at(i));
        count++;
    }

    for(i=0; i<CITY_NUM; i++){
        this->distance_arr[i][i] = 0.0;
        for(j=i+1; j<CITY_NUM; j++){
            dx = city_info[i][0] - city_info[j][0];
            dy = city_info[i][1] - city_info[j][1];
            dis = sqrt(dx*dx + dy*dy);
            this->distance_arr[i][j] = dis;
            this->distance_arr[j][i] = dis;
            this->pheromone_arr[i][j] = PHEROMONE_INIT;
            this->pheromone_arr[j][i] = PHEROMONE_INIT;
        }
    }

    for(i=0; i<CITY_NUM; i++) delete[] city_info[i];
    delete[] city_info;
}

AntSystem::~AntSystem(){
    delete[] this->agent;
    for(int i=0; i<CITY_NUM; i++){
        delete[] this->distance_arr[i];
        delete[] this->pheromone_arr[i];
    }
    delete[] this->distance_arr;
    delete[] this->pheromone_arr;
}


void AntSystem::generate_route(){
    int i, j, k;
    float prob, prob_sum, threshold;
    int last_city; 
    srand((unsigned)time(NULL));

    for(i=0; i<AGENT_NUM; i++){
        this->agent[i].reset_info();
        bool flag[CITY_NUM] = {false};
        flag[this->agent[i].id] = true;

        for(j=1; j<CITY_NUM; j++){
            prob = 0.0;
            prob_sum = 0.0;
            last_city = this->agent[i].last_city;

            for(k=0; k<CITY_NUM; k++){
                if(flag[k] == false){
                    prob_sum += pow(this->pheromone_arr[last_city][k], ALPHA) * pow(1.0 / this->distance_arr[last_city][k], BETA);
                }
            }

            threshold = float(rand()) / RAND_MAX;

            for(k=0; k<CITY_NUM; k++){
                if(flag[k] == false){
                    prob += pow(this->pheromone_arr[last_city][k], ALPHA) * pow(1.0 / this->distance_arr[last_city][k], BETA) / prob_sum;
                    if(prob > threshold){
                        this->agent[i].register_city(k);
                        flag[k] = true;
                        break;                
                    }
                }
            }
        }
    }
}

void AntSystem::calculate_distance(){
    int i, j, base, comp;
    float distance, best;

    for(i=0; i<AGENT_NUM; i++){
        distance = 0.0;
        for(j=0; j<CITY_NUM; j++){
            base = this->agent[i].route[j];
            if (j == CITY_NUM - 1){
                comp = this->agent[i].route[0];
                distance += this->distance_arr[base][comp];
            }
            else{
                comp = this->agent[i].route[j+1];
                distance += this->distance_arr[base][comp];
            }
        }
        this->agent[i].set_length(distance);
        if (i==0)best = distance;
        else{
            if(best > distance) best = distance;
        }
    }
    if(this->iteration == 0){
        this->best = best;
        this->iteration = 1;
    }
    else{
        if (this->best > best) this->best = best;
    }
    cout << "current best : " << best<< "\t all best: " << this->best << endl;
}

void AntSystem::update_pheromone(){
    int i, j, base, comp;
    float add_pheromone;
    
    for(i=0; i<CITY_NUM; i++){
        for(j=i+1; j<CITY_NUM; j++){
            this->pheromone_arr[i][j] = this->pheromone_arr[i][j] * RHO;
            this->pheromone_arr[j][i] = this->pheromone_arr[i][j];
        }
    }

    for(i=0; i<AGENT_NUM; i++){
        add_pheromone = 1.0 / this->agent[i].get_length();
        for(j=0; j<CITY_NUM; j++){
            base = this->agent[i].route[j];
            if (j == CITY_NUM - 1){
                comp = this->agent[i].route[0];
            }
            else{
                comp = this->agent[i].route[j+1];
            }
            this->pheromone_arr[base][comp] += add_pheromone;
            this->pheromone_arr[comp][base] += add_pheromone;
        }
    }
}
